pkg/kv/kvserver/replica_destroy.go

```diff
@@ -227,12 +227,9 @@ func (r *Replica) setTombstoneKey(
 ) error {
 	r.mu.Lock()
 	nextReplicaID := r.mu.state.Desc.NextReplicaID
-	if nextReplicaID < externalNextReplicaID {
+	if externalNextReplicaID > nextReplicaID {
```

Comment:
Could we be stricter here and simply require that `externalNextReplicaID` is ahead of ours?
Writing a tombstone that "affects" the current replica can only happen when the replica is currently being removed. And trying to write a tombstone that doesn't affect the current replica is unnecessary but I think indicates a logic error:

I see that we have broadly two callers into `setTombstoneKey`:

- the replica is learning that it is removed because we observe a RangeDescriptor that no longer contains the current replica (either directly or maybe indirectly, via a raft message addressing the local store under a newer replicaID).
- we're getting merged away - the NextReplicaID is `MaxInt32` in that case.

The second case clearly doesn't need the conditional.
In the first, we are calling `setTombstoneKey` as part of removing the current replica, so the tombstone *definitely* needs to be larger than the replicaID.

Reply:
The reason why I'm picking this fight (btw doesn't mean that you have to fight it here) is that this method is in the snapshot-apply path[^1]. We ultimately need/want all engine mutations to live in library code, meaning that `setTombstoneKey` cannot live on `*Replica`. This doesn't mean we can't check anything (just that we'll need to refactor the checks to live elsewhere), and that it's worth understanding exactly what the invariants are.

[^1]: https://github.com/cockroachdb/cockroach/pull/93904